semanticdb/metap/src/main/scala/scala/meta/internal/metap/Main.scala

```diff
@@ -1,43 +1,100 @@
 package scala.meta.internal.metap
 
+import java.io.InputStream
 import java.io.PrintStream
 import java.nio.file._
+import java.nio.file.attribute.BasicFileAttributes
+import java.util.Comparator
 import java.util.WeakHashMap
+import java.util.jar.JarEntry
+import java.util.jar.JarFile
+
 import scala.collection.{immutable, mutable}
-import scala.compat.Platform.EOL
 import scala.math.Ordering
 import scala.util.control.NonFatal
 import scala.meta.internal.semanticdb3._
-import Diagnostic._, Severity._
-import SymbolInformation._, Kind._, Property._
-import SymbolOccurrence._, Role._
-import Type.Tag._, SingletonType.Tag._, Accessibility.Tag._
+import Diagnostic._
+import Severity._
+import SymbolInformation._
+import Kind._
+import Property._
+import SymbolOccurrence._
+import Role._
+import Type.Tag._
+import SingletonType.Tag._
+import Accessibility.Tag._
+
+import scala.collection.mutable.ArrayBuffer
 
 class Main(settings: Settings, out: PrintStream, err: PrintStream) {
   def process(): Int = {
+
     var failed = false
-    settings.paths.zipWithIndex.foreach {
-      case (arg, i) =>
-        try {
-          val stream = Files.newInputStream(Paths.get(arg))
-          try {
-            if (i != 0) out.println("")
-            val documents = TextDocuments.parseFrom(stream)
-            if (settings.format.isProto) {
-              out.println(documents.toProtoString)
-            } else {
-              documents.documents.foreach(pprint)
-            }
-          } finally {
-            stream.close()
+    var first = true
+    def processPath(path: Path, stream: InputStream): Unit = {
+      if (first) {
+        first = false
+      } else {
+        out.println("")
+      }
+      try {
+        val documents = TextDocuments.parseFrom(stream)
+        if (settings.format.isProto) {
+          out.println(documents.toProtoString)
+        } else {
+          documents.documents.foreach(pprint)
+        }
+      } catch {
+        case NonFatal(ex) =>
+          out.println(s"error: can't decompile $path")
+          ex.printStackTrace(out)
+          failed = true
+      } finally {
+        stream.close()
+      }
+    }
+
+    settings.paths.foreach { path =>
+      if (path.getFileName.toString.endsWith(".jar")) {
+        // Can't use nio.Files.walk because nio.FileSystems is not supported on Scala Native.
+        val jarfile = new JarFile(path.toFile)
+        val buf = ArrayBuffer.empty[JarEntry]
+        val entries = jarfile.entries()
+        while (entries.hasMoreElements) {
+          val entry = entries.nextElement()
+          if (entry.getName.endsWith(".semanticdb")) {
+            buf += entry
           }
-        } catch {
-          case NonFatal(ex) =>
-            out.println(s"error: can't decompile $arg")
-            ex.printStackTrace(out)
-            failed = true
         }
+        buf.sortBy(_.getName).foreach { entry =>
+          processPath(Paths.get(entry.getName), jarfile.getInputStream(entry))
```

Comment:
Does `Paths.get(entry.getName)` include the underlying `path`?

Reply:
I haven't checked, this value is only used for reporting errors on failure.